app/routes/devices.$ip_addr.state.tsx

```diff
@@ -0,0 +1,36 @@
+import { createClient } from "@hey-api/client-fetch";
+import { LoaderFunctionArgs, json } from "@remix-run/node";
+import { Link, useLoaderData, useParams } from "@remix-run/react";
+import * as Evolver from "client/services.gen";
+import { VialGrid } from "~/components/VialGrid";
+import { ENV } from "~/utils/env.server";
+
+export const handle = {
+  breadcrumb: ({
+    params,
+  }: {
+    params: { ip_addr: string; hardware_name: string };
+  }) => {
+    const { ip_addr } = params;
+    return <Link to={`/devices/${ip_addr}/state`}>state</Link>;
+  },
+};
+
+export async function loader({ params }: LoaderFunctionArgs) {
+  const { ip_addr } = params;
+  const evolverClient = createClient({
+    baseUrl: `http://${ip_addr}:${ENV.DEFAULT_DEVICE_PORT}`,
+  });
+  const { data } = await Evolver.getStateStateGet({ client: evolverClient });
+  return json({ evolverState: data });
+}
+
+export default function Hardware() {
+  const { ip_addr } = useParams();
+  const { evolverState } = useLoaderData<typeof loader>();
+  return (
+    <div>
+      <VialGrid stateData={evolverState.state} ip={ip_addr} />
```

Comment:
The vialCount could technically come from the evolver config, do you have access to such via global state or something?

Reply:
For complete context, goal is to have a view of the evolver with active vial slots, displayed in their index positions relative to inactive slots.

Right now I cant use the config to infer, vialCount, or the general layout, irl of the vials. For example - for an evolver with hardware running on vials 0,2, all i can figure out from the config is that there are 2 vials, and possibly a "slot" at index 1, that's empty. 

What i'd like to do is represent all slots on a grid that matches the hardware layout. To do this the config would have to include some properties it doesn't currently have, at a minimum a "max_index", or even better x,y dimensions from the grid.

Reply:
When someone is looking in the ui at data for vial identified as "vial 5", by the config response. We must have some way for them to know that it's the vial on the second row, in the second position, that's what the `5` refers to.

```
[
  [0,1,2,3],
  [4,5 <- this one ,6,7],
  [8,9,10,11],
  [12,13,14,15]
]
```

right now the `vials` top level property for the evolver in your office is:

```
[0,5,10,15]
```

Which is not enough info to make the required association. Must have max_index and dimensions.

All that to say - yeah - 2d array makes sense to me. Just in case it's not a typo, in your example, replace the 1 with 5:

```
[[0, None, None, None],
 [None, 5, None, None],
...
```

Reply:
I can see this needs more in the back to support the front better and some decisions to make there, so I think what you have works for now, I've created #153 to track this.

Reply:
yeah makes sense. I can definitely lift the `16` default into a global variable to encapsulate it. If I come across similar things I'll also hoist them - altogether to make sure it's clear the front end is making assumptions about the state of the world outside of what the ng service is describing.